File: 00_SHARE/01_past_questions/01_QE/22H1/22H1_Q2.cpp
#include <iostream>
#include <string>
#include <vector>
#include <queue>
#include <unordered_map>

using namespace std;

class GNode {
public:
    string id;
    string color;
    int distance;
    GNode* parent;

    GNode(string _id, string _color = "W", int _d = 0, GNode* _p = nullptr)
        : id(_id), color(_color), distance(_d), parent(_p) {}

    string getId() const {
        return id;
    }
};

unordered_map<string, vector<string>> show_GNode(const unordered_map<string, vector<GNode*>>& G) {
    unordered_map<string, vector<string>> res;
    for (const auto& entry : G) {
        string key_id = entry.first;
        const vector<GNode*>& items = entry.second;
        vector<string> items_id_list;
        for (const GNode* item : items) {
            items_id_list.push_back(item->getId());
        }
        res[key_id] = items_id_list;
    }
    return res;
}

void bfs(unordered_map<string, vector<GNode*>>& G, GNode* s) {
    queue<GNode*> que;
    s->color = "G";

    que.push(s);

    while (!que.empty()) {
        GNode* node = que.front();
        que.pop();

        for (GNode* neigh : G[node->id]) {
            if (neigh->color == "W") {
                neigh->color = "G";
                neigh->distance = node->distance + 1;
                neigh->parent = node;
                que.push(neigh);
            }
        }

        node->color = "B";
    }
}

vector<vector<string>> level_partition(unordered_map<string, vector<GNode*>>& G, GNode* s) {
    unordered_map<int, vector<string>> level_dict;
    vector<vector<string>> res;

    bfs(G, s);

    for (const auto& entry : G) {
        const string key = entry.first;
        if (level_dict.find(G[key][0]->distance) != level_dict.end()) {
            level_dict[G[key][0]->distance].push_back(key);
        } else {
            level_dict[G[key][0]->distance] = {key};
        }
    }

    for (int i = 0; i < level_dict.size(); i++) {
        res.push_back(level_dict[i]);
    }

    return res;
}

int main() {
    GNode r("r"), s("s"), t("t"), u("u"), v("v");
    GNode w("w"), x("x"), y("y");

    unordered_map<string, vector<GNode*>> G;
    G["r"] = {&s, &v};
    G["s"] = {&w, &r};
    G["t"] = {&w, &x, &u};
    G["u"] = {&t, &x, &y};
    G["v"] = {&r};
    G["w"] = {&s, &t, &x};
    G["x"] = {&w, &t, &u, &y};
    G["y"] = {&x, &u};

    unordered_map<string, vector<string>> result = show_GNode(G);
    for (const auto& entry : result) {
        cout << entry.first << ": ";
        for (const string& item : entry.second) {
            cout << item << " ";
        }
        cout << endl;
    }

    vector<vector<string>> levels = level_partition(G, &s);
    for (int i = 0; i < levels.size(); i++) {
        cout << "Level " << i << ": ";
        for (const string& node : levels[i]) {
            cout << node << " ";
        }
        cout << endl;
    }

    return 0;
}

File: 00_SHARE/01_past_questions/01_QE/22H1/22H1_Q2.py
## Q2 ##
class GNode:
    def __init__(self, id, color="W", d=0, p=None):
        '''
        [color spec]
        W: not visited
        G: visited but its neighbors are not visited 
        B: visited and all the neighbors are visited 
        '''
        self.id = id        # id is a string
        self.color = color  # color (status) of node
        self.distance = d
        self.parent = p
  
    def __str__(self):
        return self.id
  
## utils
def show_GNode(G: dict) -> dict:
    #0. set params 
    res = {}
    for key in G.keys():
        key_id = key.id
        items_id_list = []
        for item in G[key]:
            items_id_list.append(item.id)
        # print(key_id)
        # print(items_id_list)
        res[key_id] = items_id_list
        # print(res)
    
    return res

### (a)
def bfs(G: dict, s: GNode) -> None:
    queue = [s]
    s.color = "G"

    while queue:
        node = queue.pop(0)
        # print(node.id)
        for neigh in G[node]:
            if neigh.color == "W":
                neigh.color = "G"
                neigh.distance = node.distance + 1
                neigh.parent = node
                queue.append(neigh)
        node.color = "B"

    return None

### (b)
def level_partition(G: dict,s: GNode) -> list:

    #0. set params 
    level_dict = {}
    res = []
    
    #1. do bfs 
    bfs(G, s)

    #2. add distance
    for key in G.keys():
        # print(level_dict)
        if key.distance in level_dict:
            level_dict[key.distance].append(key.id)
        else:
            level_dict[key.distance] = [key.id]

    #3. change into list type
    for i in range(len(level_dict)):
        res.append(level_dict[i])
    
    return res

# Driver Code
if __name__ == '__main__':

    r, s, t, u, v = GNode('r'), GNode('s'), GNode('t'), GNode('u'), GNode('v')
    w, x, y = GNode('w'), GNode('x'), GNode('y')
    G = dict()
    G[r], G[s], G[t], G[u], G[v] = [s, v], [w, r], [w, x, u], [t, x, y], [r]
    G[w], G[x], G[y] = [s, t, x], [w, t, u, y], [x, u]

    print(show_GNode(G))
    # print(bfs(G, s))
    print(level_partition(G, s)) #[[s], [r, w] ,[v, t, x], [u, y]]
